// @npmcli/arborist/lib/arborist/index.js
// The arborist manages three trees:
// - actual
// - virtual
// - ideal
//
// The actual tree is what's present on disk in the node_modules tree
// and elsewhere that links may extend.
//
// The virtual tree is loaded from metadata (package.json and lock files).
//
// The ideal tree is what we WANT that actual tree to become.  This starts
// with the virtual tree, and then applies the options requesting
// add/remove/update actions.
//
// To reify a tree, we calculate a diff between the ideal and actual trees,
// and then turn the actual tree into the ideal tree by taking the actions
// required.  At the end of the reification process, the actualTree is
// updated to reflect the changes.
//
// Each tree has an Inventory at the root.  Shrinkwrap is tracked by Arborist
// instance.  It always refers to the actual tree, but is updated (and written
// to disk) on reification.

// Each of the mixin "classes" adds functionality, but are not dependent on
// constructor call order.  So, we just load them in an array, and build up
// the base class, so that the overall voltron class is easier to test and
// cover, and separation of concerns can be maintained.

const {resolve} = require('path')
const {homedir} = require('os')

const mixins = [
  require('../tracker.js'),
  require('./pruner.js'), // 修剪
  require('./deduper.js'), // 去重
  require('./audit.js'), // 审核
  require('./build-ideal-tree.js'), // 构建理想目录树
  require('./load-workspaces.js'), // 加载工作区
  require('./load-actual.js'), // 加载当前实际树
  require('./load-virtual.js'), // 加载虚拟树
  require('./rebuild.js'), // 重构
  require('./reify.js'), // 依赖包整理
]

// ** 函数 compose 思路，实现了多个 class 的组合式继承，且初始继承对象为 events
const Base = mixins.reduce((a, b) => b(a), require('events'))

class Arborist extends Base {
  constructor (options = {}) {
    process.emit('time', 'arborist:ctor')
    super(options)
    this.options = {
      nodeVersion: process.version,
      ...options,
      path: options.path || '.',
      cache: options.cache || `${homedir()}/.npm/_cacache`,
    }
    this.cache = resolve(this.options.cache)
    this.path = resolve(this.options.path)
    process.emit('timeEnd', 'arborist:ctor')
  }
}

module.exports = Arborist
